clippy_lints/src/loops/never_loop.rs

```diff
@@ -170,3 +198,27 @@ fn never_loop_expr_branch<'a, T: Iterator<Item = &'a Expr<'a>>>(e: &mut T, main_
     e.map(|e| never_loop_expr(e, main_loop_id))
         .fold(NeverLoopResult::AlwaysBreak, combine_branches)
 }
+
+fn for_to_if_let_sugg(
+    cx: &LateContext<'_>,
+    loop_expr: &Expr<'_>,
+    iterator: &Expr<'_>,
+    pat: &Pat<'_>,
+) -> Option<String> {
+    // Check whether the iterator expr needs parens to parse as `(iterator).into_iter()`.
+    // Ranges need special casing because they are converted to a struct in HIR.
+    let (open_paren, close_paren) = if iterator.precedence().order() < PREC_POSTFIX || is_range_literal(iterator) {
+        ("(", ")")
+    } else {
+        ("", "")
+    };
+    let iterator_snippet = snippet_opt(cx, iterator.span)?;
+    let pat_snippet = snippet(cx, pat.span, "_");
+
+    #[rustfmt::skip]
+    let sugg = format!(
+"if let Some({pat}) = {open_paren}{iterator}{close_paren}.into_iter().next() {{
+    /* ... */
+}}", pat = pat_snippet, open_paren = open_paren, iterator = iterator_snippet, close_paren = close_paren);
+    Some(reindent_multiline(sugg.into(), true, indent_of(cx, loop_expr.span)).into())
```

Comment:
Instead of having `{ /* ... */ }` in the suggestion, can you try just replacing the "head" of the loop block? Then I think you can do something like `span_label(entire_loop_span, "in this loop")` to show more context in the output.

Reply:
Then the suggestion won't compile if there are any `break`s or `continue` in the original loop. Is that ok?

Reply:
Reachable continues can't be there (because it's a never loop). I think a fallback to your original `{ /* ... */ }` can be good in precedence of `break`.

Reply:
It's okay to make a suggestion that would require additional changes, just use `Applicability::Unspecified`. You could additionally add a note for the user pointing out `break`s but I don't think it's necessary.